rfcs/xxxx-collage.md

```diff
@@ -0,0 +1,863 @@
+  dependencies or incompatibilities and may attemt to mix candidate kernels we can't support in prod. It's also possible
+  for two BYOC integrations to have incompatible runtimes.
+- **Additive kernel cost assumption**: Collage as per this design assumes the cost of running candidate kernels is
+  additive, plus a small launch penalty. However cache effects can dominate measured latency, particularly for 'light'
+  kernels. Thus there may be a **additive error** in the final result:
+
+  > additive_error = measured_latency(collage_partitioning) - sum_{kernel} (estimated_latency(kernel) + penalty)
+
+  The evolutionary search explored by the Collage paper can help here since it uses measured end-to-end model latency as
+  its cost function, but we're deferring that to future work.
+
+- **Limited search space**: Naively exploring all sub-graphs is O(n!), so we need to constrain the search. The easiest
+  approach is just to limit candidate kernels to sub-graphs of just a few operators. This can mean significatly faster
+  candidates are not explored, yielding a partitioning with high **optimality loss**:
+
+  > optimality_loss = measured_latency(collage_partitioning) - measured_latency(true_optimal_partitioning)
+
+  Though the 'true' optimal partitioning may be infeasible to find, the user may easily discover a high
+  **apparent loss**, eg by comparing the Collage result with a traditional BYOC partitioning result:
+
+  > apparent_loss = measured_latency(collage_partitioning) - measured_latency(users_own_partitioning)
+
+- **Fragile toolchains**: Some BYOC toolchains are intended to be stand-alone compilers in their own right, and have
+  been tuned against common models and include global flags to guide optimizations such as reducing precision. However
+  Collage will only feed these toolchains smaller sub-graphs, thus making the limited search space problem more severe.
+- **High variance in lightweight kernels**: Small kernels can have high variance, thus the choice of which toolchain to
+  use can be arbitrary. We probably want to i) validate our variance estimator is accurate, ii) choose a percentile
+  slightly above 50% for the estimated candidate kernel latency, and iii) fall back to hard-coded priorities when the
+  measured variance is too high.
+- **Repeated FuseOps**: Some passes (eg `ManifestAlloc`) introduce new calls to primitive function which must be fused
+  and lowered, even though the main work of fusion and lowering has already occurred. We'll need to either
+  retain `FuseOps`, or ensure `CollageFuseOps` retains the efficiency and handling of `FuseOps` when there's no
+  toolchain ambiguity.
+- **Explainability**: It's easy to show the user the final partitioning and estimated times for each kernel, but harder
+  to show why that partitioning won out over all others during search.
+- **Does not subsume `partition_for_<toolchain>`**: We don't have any plan to deprecate the existing patterns of each
```

Comment:
I'm a bit stuck on what do about the global rewrite passes many of these BYOC integrations want to include. So I was thinking the UMA effort can solves the API aspect (the user does not need to know about all the ad-doc functions they need to invoke for their intended target) without worrying about the search/turning aspect, Collage can support search/tuning but possibly only for 'well behaved' BYOC integrations, and once both are stable we see where we stand about the UMA partitioning (ie call the usual passes exactly as done today) being subsumed by the Collage partitioning.

Hope that makes sense.

Reply:
Yeah, the pre-partitioning passes are a real problem. There are two main reasons for these:

1. Simplifying pattern matching e.g. running FoldScaleAxis to handle BatchNorm fusion I assume these would be run before Collage Partitioning
2. Layout transforms, these need to be run pre partitioning for both PatternMatching and to avoid Layout transforms inside the Matched Subgraphs.

Reply:
I added a rough section on 'Pass Ordering' to try to get my head around the constraints. I've been pretty focused on a very narrow set of BYOC & library targets so this whole issue caught me off guard.

I think if passes like FoldScaleAxis can be considered 'Safe global' (using my made-up taxonomy from that section) then we can agree to move them into the standard prefix to run before collage. However, in the TVM flow that pass is currently run quite late. So for all of these passes we'd need to:
 - decide all BYOC patterns will be written w.r.t. the canonical form enforced by the pass
 - fix any BYOC patterns which break because of that change
 - check the pass can moved backwards over other passes we wish to leave on the TVM-only side
 I'd been hoping I could stay out of that game.

Another approach is to generalize the patterns so they are pre-canonicalization w.r.t. those passes, then move the passes into the lowering function (on the assumption they are needed downstream). 

And finally, a third approach is to just declare that only some BYOC targets are 'collage ready'. I think that's where we've been all along with the prototyping but just didn't realize it. :-)

Anyway, I'm going to switch back to prototyping mode then come back to the RFC with what I've learned.

Reply:
I ended up covering this in Appendix C.

For the 'I just want to partition for BYOC x' case I think that's where UMA can take away all the manual calls to ad-hoc functions.
Eventually UMA could defer to Collage to actually do the partitioning, but only after we're confident there's no seach overhead if the user want to do things manually. I think we can tackle that later.